File: client/common/byte_buffer.py
"""
Byte Buffer for Marshalling and Unmarshalling
"""

import struct


class ByteBuffer:
    """Helper class for marshalling and unmarshalling data."""
    
    def __init__(self, data: bytes = b''):
        self.buffer = bytearray(data)
        self.read_pos = 0
    
    # Write operations
    def write_uint8(self, val: int):
        """Write an 8-bit unsigned integer."""
        self.buffer.append(val & 0xFF)
    
    def write_uint16(self, val: int):
        """Write a 16-bit unsigned integer in network byte order."""
        self.buffer.extend(struct.pack('!H', val))
    
    def write_uint32(self, val: int):
        """Write a 32-bit unsigned integer in network byte order."""
        self.buffer.extend(struct.pack('!I', val))
    
    def write_time(self, val: int):
        """Write a time_t value (as 32-bit unsigned int)."""
        self.write_uint32(val)
    
    def write_string(self, s: str):
        """Write a length-prefixed string."""
        encoded = s.encode('utf-8')
        self.write_uint16(len(encoded))
        self.buffer.extend(encoded)
    
    # Read operations
    def read_uint8(self) -> int:
        """Read an 8-bit unsigned integer."""
        if self.read_pos >= len(self.buffer):
            raise ValueError("Buffer underflow")
        val = self.buffer[self.read_pos]
        self.read_pos += 1
        return val
    
    def read_uint16(self) -> int:
        """Read a 16-bit unsigned integer from network byte order."""
        if self.read_pos + 2 > len(self.buffer):
            raise ValueError("Buffer underflow")
        val = struct.unpack('!H', self.buffer[self.read_pos:self.read_pos + 2])[0]
        self.read_pos += 2
        return val
    
    def read_uint32(self) -> int:
        """Read a 32-bit unsigned integer from network byte order."""
        if self.read_pos + 4 > len(self.buffer):
            raise ValueError("Buffer underflow")
        val = struct.unpack('!I', self.buffer[self.read_pos:self.read_pos + 4])[0]
        self.read_pos += 4
        return val
    
    def read_time(self) -> int:
        """Read a time_t value (as 32-bit unsigned int)."""
        return self.read_uint32()
    
    def read_string(self) -> str:
        """Read a length-prefixed string."""
        length = self.read_uint16()
        if self.read_pos + length > len(self.buffer):
            raise ValueError("Buffer underflow")
        s = self.buffer[self.read_pos:self.read_pos + length].decode('utf-8')
        self.read_pos += length
        return s
    
    def get_data(self) -> bytes:
        """Get the complete buffer as bytes."""
        return bytes(self.buffer)
    
    def remaining(self) -> int:
        """Get the number of unread bytes."""
        return len(self.buffer) - self.read_pos

File: java_client/src/main/java/com/facility/common/ByteBuffer.java
package com.facility.common;

import java.io.ByteArrayOutputStream;
import java.nio.charset.StandardCharsets;

/**
 * Byte Buffer for Marshalling and Unmarshalling
 */
public class ByteBuffer {
    private ByteArrayOutputStream buffer;
    private byte[] readBuffer;
    private int readPos;

    public ByteBuffer() {
        this.buffer = new ByteArrayOutputStream();
        this.readBuffer = new byte[0];
        this.readPos = 0;
    }

    public ByteBuffer(byte[] data) {
        this.buffer = new ByteArrayOutputStream();
        this.readBuffer = data;
        this.readPos = 0;
    }

    // Write operations
    public void writeUInt8(int val) {
        buffer.write(val & 0xFF);
    }

    public void writeUInt16(int val) {
        buffer.write((val >> 8) & 0xFF);
        buffer.write(val & 0xFF);
    }

    public void writeUInt32(long val) {
        buffer.write((int)((val >> 24) & 0xFF));
        buffer.write((int)((val >> 16) & 0xFF));
        buffer.write((int)((val >> 8) & 0xFF));
        buffer.write((int)(val & 0xFF));
    }

    public void writeTime(long val) {
        writeUInt32(val);
    }

    public void writeString(String s) {
        byte[] encoded = s.getBytes(StandardCharsets.UTF_8);
        writeUInt16(encoded.length);
        buffer.write(encoded, 0, encoded.length);
    }

    // Read operations
    public int readUInt8() {
        if (readPos >= readBuffer.length) {
            throw new RuntimeException("Buffer underflow");
        }
        return readBuffer[readPos++] & 0xFF;
    }

    public int readUInt16() {
        if (readPos + 2 > readBuffer.length) {
            throw new RuntimeException("Buffer underflow");
        }
        int val = ((readBuffer[readPos] & 0xFF) << 8) |
                  (readBuffer[readPos + 1] & 0xFF);
        readPos += 2;
        return val;
    }

    public long readUInt32() {
        if (readPos + 4 > readBuffer.length) {
            throw new RuntimeException("Buffer underflow");
        }
        long val = ((long)(readBuffer[readPos] & 0xFF) << 24) |
                   ((long)(readBuffer[readPos + 1] & 0xFF) << 16) |
                   ((long)(readBuffer[readPos + 2] & 0xFF) << 8) |
                   ((long)(readBuffer[readPos + 3] & 0xFF));
        readPos += 4;
        return val;
    }

    public long readTime() {
        return readUInt32();
    }

    public String readString() {
        int length = readUInt16();
        if (readPos + length > readBuffer.length) {
            throw new RuntimeException("Buffer underflow");
        }
        String s = new String(readBuffer, readPos, length, StandardCharsets.UTF_8);
        readPos += length;
        return s;
    }

    public byte[] getData() {
        return buffer.toByteArray();
    }

    public int remaining() {
        return readBuffer.length - readPos;
    }

    public int size() {
        return buffer.size();
    }

    public void write(byte[] data) {
        buffer.write(data, 0, data.length);
    }
}
